Apollo4 adc driver

RichardSWheatley opened the pull request:
Update Ambiq ADC Driver to add apollo4 plus

Signed-off-by: Richard Wheatley <richard.wheatley@ambiq.com>

RichardSWheatley commented:
Converted to draft until https://github.com/zephyrproject-rtos/zephyr/pull/73763 is finished.
Then I can rebase and add my changes here.

RichardSWheatley commented:
@decsny please see my comments on https://github.com/zephyrproject-rtos/zephyr/pull/73763 addressing your concerns there.

RichardSWheatley commented:
Updated due to clang-format request from Zephyr-bot.

RichardSWheatley commented:
Added ASYNC for apollo3/4 and PM in for apollo4 plus

decsny commented on drivers/adc/adc_ambiq.c:
```diff
@@ -343,47 +384,56 @@ static int adc_ambiq_pm_action(const struct device *dev, enum pm_device_action a
-	{                                                                                      \
+#define ADC_AMBIQ_INIT(n)                                                                          \
+	PINCTRL_DT_INST_DEFINE(n);                                                                 \
+	ADC_AMBIQ_DRIVER_API(n);                                                                   \
```
please do format updates in a separate commit

RichardSWheatley replied:
I will fix tomorrow. Thank you!

RichardSWheatley committed:
drivers: adc: update ambiq adc driver

Update Ambiq adc driver to add apollo4
fixed typ of BUMBER to NUMBER
Move buffer size prio to configuring slots
and enabling the ADC
removed #if defined(CONFIG_PM_DEVICE_RUNTIME)

Signed-off-by: Richard Wheatley <richard.wheatley@ambiq.com>

RichardSWheatley committed:
dts: arm: ambiq: add ambiq adc to dtsi file

Add Ambiq adc to apollo4 DTSI

Signed-off-by: Richard Wheatley <richard.wheatley@ambiq.com>

RichardSWheatley committed:
boards: ambiq: add adc to apollo4p boards

Add adc to apollo4 plus boards

Signed-off-by: Richard Wheatley <richard.wheatley@ambiq.com>

RichardSWheatley committed:
tests: drivers: adc: adc_api: boards: add conf file

Add configuration file to adc_api test and adc sample

Signed-off-by: Richard Wheatley <richard.wheatley@ambiq.com>

RichardSWheatley committed:
drivers: adc: format ambiq adc driver

run clang-format on ambiq adc driver

Signed-off-by: Richard Wheatley <richard.wheatley@ambiq.com>

decsny commented on drivers/adc/adc_ambiq.c:
the commit message on eda385da95c5b3967b1e685a5e6a51820071f505 should be improved, don't link to a github "conversation", instead, clearly list all the things the commit is doing and why

decsny commented on drivers/adc/adc_ambiq.c:
```diff
@@ -157,9 +162,15 @@ static int adc_ambiq_start_read(const struct device *dev, const struct adc_seque
 		return -EINVAL;
 	}
+
+	error = adc_ambiq_check_buffer_size(sequence, active_channels);
```
for example, why has this been moved

decsny reviewed (commented):
to me the last 4 commits seem fine, just the first commit to the driver please make more clear in the message what it does

RichardSWheatley commented:
Can anyone tell me why the twister talent failed with the same code it passed with for the last week?

aaronyegx commented:
> Can anyone tell me why the twister talent failed with the same code it passed with for the last week?

See the errors:

/__w/zephyr/zephyr/drivers/serial/uart_pl011_ambiq.h:74:12: error: 'uart_ambiq_pm_action' defined but not used [-Werror=unused-function]
   74 | static int uart_ambiq_pm_action(const struct device *dev, enum pm_device_action action)
      |            ^~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors

It seems the uart_ambiq_pm_action() is only used by Apollo3, which includes in https://github.com/zephyrproject-rtos/zephyr/pull/77110

RichardSWheatley commented:
> > Can anyone tell me why the twister talent failed with the same code it passed with for the last week?
> 
> 
> 
> See the errors:
> 
> 
> 
> /__w/zephyr/zephyr/drivers/serial/uart_pl011_ambiq.h:74:12: error: 'uart_ambiq_pm_action' defined but not used [-Werror=unused-function]
> 
>    74 | static int uart_ambiq_pm_action(const struct device *dev, enum pm_device_action action)
> 
>       |            ^~~~~~~~~~~~~~~~~~~~
> 
> cc1: all warnings being treated as errors
> 
> 
> 
> It seems the uart_ambiq_pm_action() is only used by Apollo3, which includes in https://github.com/zephyrproject-rtos/zephyr/pull/77110

So, only the apollo3 UART PM was addresssed in your update?

I will have to address that first.

RichardSWheatley commented:
> > Can anyone tell me why the twister talent failed with the same code it passed with for the last week?
> 
> See the errors:
> 
> /__w/zephyr/zephyr/drivers/serial/uart_pl011_ambiq.h:74:12: error: 'uart_ambiq_pm_action' defined but not used [-Werror=unused-function] 74 | static int uart_ambiq_pm_action(const struct device *dev, enum pm_device_action action) | ^~~~~~~~~~~~~~~~~~~~ cc1: all warnings being treated as errors
> 
> It seems the uart_ambiq_pm_action() is only used by Apollo3, which includes in #77110

@decsny Would I be able to fix this issue in this PR or will I need to create another?

RichardSWheatley commented:
This is now blocked by: https://github.com/zephyrproject-rtos/zephyr/pull/77477
The fix for the UART error created when apollo3 UART PM update was merged.
